### PR TITLE
feat: add requiresSyncData and verify-sync test

### LINKED ISSUE #248: New Annotation for Constraint Templates' which require SyncData
I propose (and can submit the PR) for adding a new Constraint Templates annotation for the 4 Constraint Templates which require SyncData:
- `K8sPodDisruptionBudget`
- `K8sStorageClass`
- `K8sUniqueIngressHost`
- `K8sUniqueServiceSelector`

Here is an example of what that would look like for `K8sStorageClass`:

```
  annotations:
    metadata.gatekeeper.sh/requiresSyncData: |-
      [
         {
            "group":"storage.k8s.io",
            "version":"v1",
            "kind":"StorageClass"
         }
      ]
```

For Constraint Templates that don't require SyncData, it should simply be omitted.

### PR DESCRIPTION
**Pull request by apeabody**
* Adds a new `metadata.gatekeeper.sh/requiresSyncData` annotation to templates using `data.inventory`
* Add a `sync.yaml` for `uniqueingresshost` template
* Adds a new `require-sync` test which requires templates using data.inventory have both a `sync.yaml` and `metadata.gatekeeper.sh/requiresSyncData` annotation

Fixes #248

### CODE REVIEW

**Comment by apeabody** [comment]
I left the template versions as-is given the actual rego/etc hasn't changed, but let me know if we should bump the minor or patch.  Thanks!

File: library/general/uniqueingresshost/template.yaml
Comment (maxsmythe) [inline library/general/uniqueingresshost/template.yaml:11]: Do all K8s versions have all of these GVKs? Interesting to think about cases where there may be missing GVKs on older clusters (or newer clusters if the "extensions" group has been removed)

File: scripts/require-sync.sh
Comment (maxsmythe) [inline scripts/require-sync.sh:10]: This is not a 100% reliable way of detecting requiring inventory (e.g. `input.parameters.data.inventory` would trigger this as well).

More reliable would be to attempt to compile the template with `inventory` disabled in the constraint framework, but TBH I'm not sure this is a blocker.

It might be good to give `gator test` the ability to turn of inventory (@julianKatz 's old PR), as a way of easily and reliably detecting something like this via bash.

**Review by maxsmythe — COMMENTED** [review]
Basically LGTM. Ideally we'd use more reliable detection for `data.inventory` usage, but I think requiring it would depend on how much work that entails.

The question around "what about clusters for whom these resources don't exist" is salient if we want to think about the expected uses of these annotations.

**Comment by maxsmythe** [comment]
Also, WRT template versions, leaving them as-is seems right to me (no Rego changes)

File: library/general/uniqueingresshost/template.yaml
Comment (ekitson) [reply library/general/uniqueingresshost/template.yaml:11]: Ah interesting. Maybe this schema needs a slight tweak to account for this, eg:
requiresSyncData: [list of {requirement}s, ANDed]
{requirement}: [list of {GVK}s, ORed]

That would give enough expression to account for things like "v1beta1 fully deprecated in favor of v1" or "Deployments migrated from extensions to apps"

File: scripts/require-sync.sh
Comment (apeabody) [reply scripts/require-sync.sh]: Thanks @maxsmythe, yes it would be great if `gator test` had that capability.  In the meantime I updated the grep to `' data.inventory.'` (space at the start) which will at least avoid the possibility of a false match on `input.parameters.data.inventory`.

File: library/general/uniqueingresshost/template.yaml
Comment (maxsmythe) [reply library/general/uniqueingresshost/template.yaml:11]: I wouldn't presume kind to be stable, but otherwise +1

Comment (apeabody) [reply library/general/uniqueingresshost/template.yaml]: Thanks @maxsmythe @ekitson !  Based on the feedback the PR has been updated to use this [ [{}OR{}] AND [{}OR{}] ] format:

```
    metadata.gatekeeper.sh/requiresSyncData: |
      "[
        [
          {
            "groups":["extensions"],
            "versions": ["v1beta1"],
            "kinds": ["Ingress"]
          },
          {
            "group": ["networking.k8s.io"],
            "version": ["v1beta1", "v1"],
            "kind": ["Ingress"]
          }
        ]
      ]"
```

**Review by maxsmythe — APPROVED** [review]
LGTM!

File: scripts/require-sync.sh
Comment (nilekhc) [inline scripts/require-sync.sh:1]: @maxsmythe Do we want to shift towards `Golang` for such new tasks?

Comment (maxsmythe) [reply scripts/require-sync.sh:1]: +1 good point @apeabody that should also let us alleviate the issue with detecting referential constraints.

File: library/general/uniqueingresshost/template.yaml
Comment (apeabody) [reply library/general/uniqueingresshost/template.yaml]: Both the schema and detection method have been updated.

File: scripts/require-sync.sh
Comment (apeabody) [reply scripts/require-sync.sh]: Replaced with a golang version which also has a better check for referential constraints.

File: website/docs/replicalimits.md
Comment (sozercan) [inline website/docs/replicalimits.md:138]: I think we need the revert this change

File: website/docs/replicalimits.md
Comment (sozercan) [inline website/docs/replicalimits.md:171]: same here

File: artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml
Comment (sozercan) [inline artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml:7]: @nilekhc would we want to bump the version in this case?

File: scripts/require-sync/go.mod
Comment (sozercan) [inline scripts/require-sync/go.mod:7]: do we need the otel dependencies?

File: scripts/require-sync/main.go
Comment (sozercan) [inline scripts/require-sync/main.go:183]: does this need to be moved inside the referential check?

File: .github/workflows/workflow.yaml
Comment (sozercan) [inline .github/workflows/workflow.yaml:88]: if this fails, can we output an actionable message to the user

File: website/docs/replicalimits.md
Comment (apeabody) [reply website/docs/replicalimits.md:138]: Thanks @sozercan - It appears there were manually updated in https://github.com/open-policy-agent/gatekeeper-library/commit/1def56b845b5c7a1acf37b9f611c7e7a8835dd68, and are being changed back "make generate generate-website-docs".  I'll need to look into it further.

Comment (apeabody) [reply website/docs/replicalimits.md:138]: Hi @sozercan - I traced this to an issue with the website generate script #254 that was impacting all the website policy examples.  I just opened PR #255 to resolve.

**Comment by apeabody** [comment]
Pending #255

File: .github/workflows/workflow.yaml
Comment (apeabody) [reply .github/workflows/workflow.yaml:88]: Yes, `require-sync` will return a message with the template and what is missing the user needs to add:

`template at path '%s' is missing annotation with key '%s'`
or
`sync.yaml not found in dir '%s'`

File: scripts/require-sync/main.go
Comment (apeabody) [reply scripts/require-sync/main.go]: Yes, that whole section has been rewritten.

File: scripts/require-sync/go.mod
Comment (apeabody) [reply scripts/require-sync/go.mod:7]: I believe these will be required till this issue is resolved: https://github.com/open-policy-agent/frameworks/issues/210

File: artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml
Comment (apeabody) [reply artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml:7]: I'm happy to bump if desired, however there is no change to the rego/template behavior.

File: scripts/require-sync/main.go
Comment (maxsmythe) [inline scripts/require-sync/main.go]: is this an error? Should we use `fmt.Sprintf()`?

**Review by maxsmythe — APPROVED** [review]
1 nit, but LGTM

File: scripts/require-sync/main.go
Comment (apeabody) [reply scripts/require-sync/main.go]: Hi @maxsmythe  - I believe that's from an earlier commit, it shouldn't be in the current version: https://github.com/open-policy-agent/gatekeeper-library/pull/251/files#diff-f987eb96fe4bcda2fea78c6a08551b7947c935c4a3a4d5ce8af88083957e9af2R96

File: scripts/require-sync/main.go
Comment (maxsmythe) [inline scripts/require-sync/main.go:29]: nit: `sync-file` for flag consistency

**Review by maxsmythe — APPROVED** [review]
LGTM after 1 nit

File: scripts/require-sync/main.go
Comment (apeabody) [reply scripts/require-sync/main.go]: SG - done

File: scripts/require-sync/go.mod
Comment (sozercan) [inline scripts/require-sync/go.mod:1]: can you run `go mod tidy`? i am seeing a few updates on my side and looks like `go.work.sum` file is also generated

**Review by sozercan — APPROVED** [review]
minor comment, otherwise LGTM

File: .github/workflows/workflow.yaml
Comment (ritazh) [inline .github/workflows/workflow.yaml:88]: it's good that we are checking here. should we also add validation in Gatekeeper to write an error in status if the sync resource is missing the required data in `metadata.gatekeeper.sh/requiresSyncData`?

File: scripts/require-sync/go.mod
Comment (apeabody) [reply scripts/require-sync/go.mod:1]: Done, should have done that after the rebase.

File: artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml
Comment (ritazh) [reply artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml:7]: Since we are adding a new annotation (that may be used for validation), we should at least bump the patch version IMO. WDYT?

Comment (nilekhc) [reply artifacthub/library/general/poddisruptionbudget/1.0.0/template.yaml:7]: Yup. We should bump the patch version.

File: .github/workflows/workflow.yaml
Comment (apeabody) [reply .github/workflows/workflow.yaml:88]: Thanks @ritazh! - Absolutely, however that scope of work is beyond what I can include in this particular PR.  Should we track as a feature request issue for future contribution?  @maxsmythe @ekitson

Comment (maxsmythe) [reply .github/workflows/workflow.yaml:88]: > or is missing in data.inventory

I don't think this can be reliably determined: is it missing because it's not being synced or because there is nothing to sync?

> write an error in status if the required data specified in metadata.gatekeeper.sh/requiresSyncData is not part of the sync resource

This makes sense to do, but is non-trivial. I'd file an issue in G8r.

Comment (apeabody) [reply .github/workflows/workflow.yaml:88]: Opened: https://github.com/open-policy-agent/gatekeeper/issues/2393

**Comment by apeabody** [comment]
Hi Everyone - Is anything else needed before this can be merged?  Thanks!

**Comment by maxsmythe** [comment]
@ritazh any outstanding comments?

**Review by ritazh — APPROVED** [review]
LGTM

**Pull request merged**
